### PR TITLE
handshake refactor

### LINKED ISSUE #329: Refactor connection/channel identifier handling
[address significant changes](https://github.com/cosmos/cosmos-sdk/pull/7993)

I will handle this since the changes are non-trivial

### PR DESCRIPTION
**Pull request by colin-axner**
closes: #329 

contains changes from #330, couldn't push that branch to the repo for some reason so I can't open this pr against that one

TODOs:

- [x] Parse CreateClient events to get client ids (set in PathEnd and config file)
- [x] Parse ConnOpenInit/ConnOpenTry events to get connection ids (set in PathEnd and config file)
- [x] Parse ChanOpenInit/ChanOpenTry events to get channel ids (set in PathEnd and config file)
- [ ] Validate that a user-filled config is mutually coherent. (ie the clients point to each other, the connections are built on underlying clients, etc)

### CODE REVIEW

File: relayer/chain.go
Comment (colin-axner) [inline relayer/chain.go:327]: I think it makes sense to cut down on the number of `Send` messages and just handle all the logic in one place

File: relayer/connection-tx.go
Comment (colin-axner) [inline relayer/connection-tx.go:111]: pretty sure this was a bug. The retry callback above will update the sync headers if it fails, but this go routine, which presumably runs in parallel, needs the sync header height. It is safer to just do these calls synchronously instead of async. Would prefer to have a stable > fast implementation

File: relayer/connection-tx.go
Comment (colin-axner) [inline relayer/connection-tx.go]: After talking with @jackzampolin, we want to make the default behaviour of the relayer to reuse existing clients/connection/channels instead of creating a new one by default. This might require adding additional helper functions on the SDK side. For now default will new things, but we can easily add this in later by inserting a function here (or after client creation steps are complete) to fill the potentially empty connection identifiers

**Comment by colin-axner** [comment]
I'd suggest looking at the file, the diffs are hard to read

File: relayer/connection-tx.go
Comment (AdityaSripal) [inline relayer/connection-tx.go]: This should be configurable

Comment (colin-axner) [reply relayer/connection-tx.go]: will make a function arg

File: relayer/connection-tx.go
Comment (AdityaSripal) [reply relayer/connection-tx.go]: I think this is a nice-to-have. We should get it to work by just creating new identifiers and merge this PR. Future work can implement this feature

File: relayer/channel-tx.go
Comment (colin-axner) [inline relayer/channel-tx.go]: shouldn't this be validate connection paths?

File: relayer/client-tx.go
Comment (colin-axner) [inline relayer/client-tx.go]: this isn't that bad for state based event parsing

File: relayer/chain.go
Comment (jackzampolin) [reply relayer/chain.go:327]: This is much nicer. We could also add retries for account number in here too...

File: relayer/chain.go
Comment (jackzampolin) [inline relayer/chain.go:636]: Nice, also this looks like easy to retry if the returned proofs are nil.

Comment (jackzampolin) [reply relayer/chain.go:636]: named returns would be nice here for docs

File: relayer/client-tx.go
Comment (jackzampolin) [inline relayer/client-tx.go:94]: Nice

File: relayer/client-tx.go
Comment (jackzampolin) [inline relayer/client-tx.go]: `if err := overwriteConfig(config); err != nil { return err }` should do this, but `Config` stuff is all in `cmd` package...

Comment (jackzampolin) [reply relayer/client-tx.go]: maybe we make a note on the function that the chains that are passed in have modified `PathEnd`s when they are returned?

File: relayer/client-tx.go
Comment (jackzampolin) [inline relayer/client-tx.go]: When would we hit this case?

File: relayer/ics24.go
Comment (jackzampolin) [inline relayer/ics24.go:91]: Nice, should have done this awhile ago

File: relayer/path.go
Comment (jackzampolin) [inline relayer/path.go:111]: Maybe we can pull all that `RandLowerCaseLetterString` code out...

File: relayer/relayMsgs.go
Comment (jackzampolin) [inline relayer/relayMsgs.go:141]: cc @michaelfig

**Review by jackzampolin — APPROVED** [review]
Not as bad as I feared. Looking forward to testing the changeset.

File: relayer/client-tx.go
Comment (jackzampolin) [reply relayer/client-tx.go]: Its not that bad, but its not that great.

File: relayer/relayMsgs.go
Comment (michaelfig) [reply relayer/relayMsgs.go:141]: That LGTM (the next line preserves the semantics I need).

File: relayer/client-tx.go
Comment (colin-axner) [reply relayer/client-tx.go]: If the user pre-sets the identifiers

File: relayer/client-tx.go
Comment (colin-axner) [reply relayer/client-tx.go]: This was a good suggestion, required minimal changes which was nice

File: relayer/client-tx.go
Comment (AdityaSripal) [inline relayer/client-tx.go:154]: So this should be called after the chain hangs

Comment (colin-axner) [reply relayer/client-tx.go:154]: yea, I'm a little unsure about if the correct query height is being used

File: relayer/client-tx.go
Comment (AdityaSripal) [inline relayer/client-tx.go:162]: I believe this is right
```suggestion
	height := int64(sh.GetHeight(dst.ChainID)) -1
```

but we can test live

**Comment by colin-axner** [comment]
add check that off chain light clients have been initialized before creating on chain clients

File: Makefile
Comment (AdityaSripal) [inline Makefile]: Can we change this back?

File: cmd/config.go
Comment (AdityaSripal) [inline cmd/config.go]:
```suggestion
				fmt.Printf("Skipping non chain file: %s\n", f.Name())
```

File: cmd/raw.go
Comment (AdityaSripal) [inline cmd/raw.go:328]: nit: these should be swapped

**Review by AdityaSripal — CHANGES_REQUESTED** [review]

File: cmd/raw.go
Comment (AdityaSripal) [reply cmd/raw.go:306]: Will deal with it in future PR

File: Makefile
Comment (colin-axner) [reply Makefile]: yup

File: cmd/raw.go
Comment (colin-axner) [reply cmd/raw.go:328]: the issue here is open init could succeed and open try could fail. In this case we want to update the open init id instead of creating a new id upon retry

**Comment by colin-axner** [comment]
once rc6 is cut we should bump the sdk version and merge into master

**Review by AdityaSripal — APPROVED** [review]

**Comment by colin-axner** [comment]
cc @jackzampolin I'm not sure why the builds don't work. Can we merge it or should we try to fix the builds?

This pr is currently based on an SDK commit after rc5. I'd be fine merging right now with that commit and then bumping to rc6 when that is cut, but we can also wait for rc6 to merge

**Comment by anilcse** [comment]
> cc @jackzampolin I'm not sure why the builds don't work. Can we merge it or should we try to fix the builds?
> 
> This pr is currently based on an SDK commit after rc5. I'd be fine merging right now with that commit and then bumping to rc6 when that is cut, but we can also wait for rc6 to merge

We need to update the CI to use go1.15+. Looks like that is one of the issues

**Comment by cyborgshead** [comment]
@colin-axner @jackzampolin hope this with config path support will be merged soon (now only empty follows to creation otherwise tries to query)
I tested with RC5 (demo with two chains) chains and path creation is work but there is error when trying to send tokens back cause error in coin denom (channel name channel-0)
Think it's why integration tests failed

**Comment by colin-axner** [comment]
> I tested with RC5 (demo with two chains) chains and path creation is work but there is error when trying to send tokens back cause error in coin denom (channel name channel-0)

Thanks for taking the time to test these changes. The token error is caused by a bug that was fixed after rc5 was cut. I forgot that this is why I needed to bump the SDK commit. I'm going to set the relayer on a commit that contains all the IBC fixes and then merge this pr after I fix the integration tests. We can set the relayer for rc6 once that is cut on the SDK and gaia

Edit: rc5 should be fine for the tests, we just need to update the docker image

**Comment by colin-axner** [comment]
@jackzampolin can you upload a new gaia docker image? I think that's the issue for the integration tests

**Comment by cyborgshead** [comment]
RC6 released
https://github.com/cosmos/cosmos-sdk/releases/tag/v0.40.0-rc6

**Comment by colin-axner** [comment]
I bumped to rc6, cleaned up some of my commits (squashed them down) and manually tested again that this branch still works for base functionality.

I'm going to do a regular merge (not using squash) so all the contributors to this branch get credit. I'm merging with integration tests not passing because I don't think that is a sufficient reason to block development/testing/release. We can fix any issues as they arise and fix integration tests in a follow up. I will open issues for things mentioned on this pr that should be addressed in followups

**Pull request merged**
